packages/babel-helper-replace-supers/src/index.js

```diff
@@ -26,6 +26,24 @@ function getPrototypeOfExpression(objectRef, isStatic, file, isPrivateMethod) {
   return t.callExpression(file.addHelper("getPrototypeOf"), [targetRef]);
 }
 
+function ensureClassRefValidInChildScopes(
+  path: NodePath,
+  objectRefName: string,
+) {
+  path.traverse({
+    Scopable(path) {
+      if (path.type === "ClassDeclaration" || path.type === "ClassExpression") {
+        path.skip();
+        return;
+      }
+
+      if (path.scope.hasOwnBinding(objectRefName)) {
+        path.scope.rename(objectRefName);
+      }
+    },
+  });
+}
```

Comment:
We should use `environmentVisitor` here, so that it properly handles every place where `super` refers to the current super class.

We can also avoid creating a new traversal visitor every time this function is called, to slightly improve perf:
```js
const unshadowSuperBindingVisitor = traverse.visitors.merge([
  environmentVisitor,
  {
    Scopable(path, { refName }) {
      if (path.scope.hasOwnBinding(refName)) {
        path.scope.rename(refName);
      }
    },
  },
]);
```

And then use it in `replace()` as
```js
path.traverse(unshadowSuperBindingVisitor, { refName: this.opts.refToPreserve }); 
```

Reply:
It's strange that bindings are not updated correctly in `unshadowSuperBindingVisitor`.

https://github.com/Zzzen/babel/pull/1/files#r554520633